### PR TITLE
Make Gemma4 MoE export faithful and practical

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Gemma4 MoE checkpoints (e.g. [`google/gemma-4-26B-A4B-it`](https://huggingface.co/google/gemma-4-26B-A4B-it)) were exported with several numerical and practicality defects in the MoE block. This PR fixes all of them so the export matches HuggingFace and runs on released onnxruntime.

### Fused path (`com.microsoft::MoE`)
- **Activation**: use `activation_alpha=1.702` (gelu-approx) instead of `1.0` (silu) for gelu-family experts. Gemma4 experts compute `act_fn(gate)*up` with `hidden_activation=gelu_pytorch_tanh`; `x*sigmoid(1.702x)` closely approximates gelu, whereas `alpha=1.0` emits silu — a wrong activation. A genuine silu expert still maps to `1.0`.
- **Per-expert scale**: fold `router.per_expert_scale` into `fc2` (`fc2[e] *= per_expert_scale[e]`). HF applies `top_k_weights *= per_expert_scale[idx]` after renormalization, but the fused op has no per-expert output scaling.

### Unfused fallback
- Apply the **exact** expert activation — `Gelu(approximate="tanh")` for gelu experts (Swish/silu otherwise) — instead of always emitting `Swish`.
- **Neutralize** `router.per_expert_scale` to ones after folding into `fc2` so the fallback (which reads it) does not double-apply the scale.
- Replace the per-expert static unroll (`O(E*K)` nodes — ~6k/layer at E=128/K=8, **183k** for a 30-layer 26B model, impractical for ORT session-init) with a **vectorized** dense evaluation: batched `Einsum` gate/up + down GEMMs masked by a `ScatterElements`'d top-K routing vector — `O(1)` nodes/layer (**~2k** total for 26B). Standard `MatMul`/`Einsum` accumulate in **fp32** on CUDA even for fp16 inputs, so this path is numerically faithful to HuggingFace, unlike the released-ORT fused MoE kernel which fp16-accumulates the expert GEMMs and loses accuracy at large hidden size.

## Verification

On `google/gemma-4-26B-A4B-it` (fp16, CUDA, H200, onnxruntime-gpu 1.29), the vectorized unfused build:
- **Deterministic greedy generation matches HF 24/24 tokens** (token-identical text).
- **L4 last-position** logits: top-1 argmax 1.0, cosine **0.99999**; all-positions cosine **0.9996**.
- Isolated MoE on real layer-0 weights matches an fp32 HF reference to cosine **0.998** (CPU and CUDA).

Root-cause detail: the released fused `com.microsoft::MoE` CUDA kernel fp16-accumulates the expert GEMMs (verified: op cosine 0.9998 at hidden=512 but ~0.97 at hidden=2816; identical in fp16/fp32 graph dtype), while standard `MatMul`/`Einsum` accumulate in fp32 → faithful. This PR's vectorized fallback makes that faithful path practical to load.

Tests: `pytest tests/build_graph_test.py src/mobius/models/gemma4_test.py -k gemma4` (57 passed); `ruff check`/`ruff format` clean.

## Related

- Consumer/schema side (onnx-genai canonical metadata for the Gemma4 pair): onnxruntime/onnx-genai#1716.
- Published real-weight example built with this branch: https://huggingface.co/justinchuby/onnx-genai-example-gemma4-26b-a4b

---

## Second change: emit Gemma4 final hidden state for borrowed-KV speculative drafters

Gemma4's custom `forward` did not honor `output_layer_indices`, so a Gemma4 target could not expose the hidden state a borrowed-KV speculative drafter (e.g. [`google/gemma-4-E2B-it-assistant`](https://huggingface.co/google/gemma-4-E2B-it-assistant)) consumes as its **folded-carry seed**. The onnx-genai speculative contract (onnxruntime/onnx-genai#1716) now requires `folded_carry_seed` to name a **real target output port**, so the target must emit it.

- When `config.output_layer_indices` is set, `Gemma4CausalLMModel.forward` returns the **post-final-norm** hidden state (the exact `lm_head` input, == HuggingFace `output_hidden_states=True` `hidden_states[-1]`) as a third tuple element; the gemma4 text task registers it as `hidden_states.{idx}`. `None`/empty keeps the legacy 2-tuple, so **existing exports are byte-identical** (adding the output taps an existing value — no new initializer).
- gemma4 exposes only the FINAL normalized hidden (not per-layer pre-norm), so the requested index is the last decoder layer (`num_hidden_layers - 1`), matching the HF `hidden_states[k+1]` convention where the last slot is the normalized output.

**Verification** on `google/gemma-4-E2B-it` (`output_layer_indices=[34]`, fp16, CUDA, H200): emitted `hidden_states.34` matches HF `hidden_states[-1]` at **cos=0.999999** (rel 0.0016); `model.onnx.data` byte-identical (weights unchanged). 39 gemma4 graph tests pass; ruff clean. Published example pair rebuilt with this output: [target](https://huggingface.co/justinchuby/onnx-genai-example-gemma4-e2b) + [assistant](https://huggingface.co/justinchuby/onnx-genai-example-gemma4-e2b-assistant) + [self-contained speculative](https://huggingface.co/justinchuby/onnx-genai-example-gemma4-e2b-speculative).

### CODE REVIEW

File: src/mobius/models/gemma4.py [hunk 1]
```diff
@@ -1607,71 +1651,51 @@ def _dispatch_moe_fallback(
...
+        gate_w = op.Slice(self.fc1_experts_weights, [0], [moe_inter], [1])  # [E, moe_inter, H]
+        up_w = op.Slice(
+            self.fc1_experts_weights, [moe_inter], [2 * moe_inter], [1]
+        )  # [E, moe_inter, H]
+
+        # Dense per-expert gate/up projections: [T, H] x [E, moe_inter, H] -> [T, E, moe_inter]
+        gate = op.Einsum(normed_flat, gate_w, equation="th,eih->tei")
+        up = op.Einsum(normed_flat, up_w, equation="th,eih->tei")
+        activated = op.Gelu(gate, approximate="tanh") if self._moe_is_gelu else op.Swish(gate)
+        inter = op.Mul(activated, up)  # [T, E, moe_inter]
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/gemma4.py:1702]: Fixed in 6213865d. The unfused fallback now applies the exact expert activation via the shared `ACT2FN` mapping — `self._moe_act_fn = get_activation(config.hidden_act)` — so bare `gelu` maps to erf GELU while only `gelu_pytorch_tanh`/`gelu_new`/`gelu_fast` use the tanh approximation (and `silu`→Swish). No more behavior drift for `hidden_act="gelu"`. Real Gemma4 MoE checkpoints declare `gelu_pytorch_tanh`, so their exports are unchanged. Added regression tests `TestGemma4MoeFallbackActivation` asserting the emitted op per `hidden_act`.

File: src/mobius/models/gemma4.py [hunk 2]
```diff
@@ -320,9 +320,33 @@ def _remap_moe_expert_weights(
...
+                    state_dict[key] = fc2 * pes.to(fc2.dtype).reshape(-1, 1, 1)
+        # per_expert_scale is now baked into fc2. Neutralize the router copy to
+        # ones so the unfused fallback path — which reads
+        # ``self.router.per_expert_scale`` and multiplies it into the routing
+        # weights — does not apply the scale a second time. The fused path never
+        # reads it. (The parameter is kept, not dropped, so weight loading still
+        # binds it.)
+        for key in list(state_dict.keys()):
+            if key.endswith(".router.per_expert_scale"):
+                state_dict[key] = torch.ones_like(state_dict[key])
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/gemma4.py]: Fixed in 6213865d. Neutralization is now scoped to the prefixes we actually folded: `_remap_moe_expert_weights` records each prefix whose `fc2_experts_weights` received the `per_expert_scale` fold, then sets only those `router.per_expert_scale` copies to ones. A partial state_dict that supplies `router.per_expert_scale` without its matching `fc2_experts_weights` is left intact — the scale is neither double-applied on the fallback nor silently dropped. Added `TestGemma4PerExpertScalePartialStateDict` covering both the fc2-absent (preserved) and fc2-present (folded + neutralized) cases.

**Comment by justinchuby** [comment]
## Follow-up: fixed `output_layer_indices` correctness (final-duck finding)

New head: `00ace9310ecf0e3248ab235a60e1929a2e76820e` (delta `6213865d..00ace931`, 3 files).

**Bug:** `Gemma4CausalLMModel.forward` returned the *same* shared `ir.Value` for every requested `output_layer_indices` entry, and the task registered each via `builder.add_output` (in-place rename) — so multiple indices collapsed onto a single output. Unlike the generic `TextModel`, Gemma4 also never rejected duplicate / out-of-range / negative / non-last indices.

**Decision — Option B (architecture-faithful restriction).** Gemma4's `forward` materializes *only* the final post-final-norm hidden state (the exact `lm_head` input == HF `output_hidden_states[-1]`); per-layer *pre*-norm hidden states are never produced. Rather than pretend general per-layer support (as generic `TextModel` offers), the config is restricted to exactly one index == `num_hidden_layers - 1`.

**Changes**
- `_validate_gemma4_output_layer_indices(config)`: rejects, in `TextModel`-consistent order, duplicates → out-of-range/negative (`0 <= i < num_hidden_layers`) → anything not exactly `[num_hidden_layers - 1]`, each with a clear error. `None`/empty → `None` (legacy 2-tuple contract preserved).
- `__init__` normalizes/validates via the helper.
- `forward` emits a **distinct `op.Identity(hidden_states)` per index**, so every graph output owns a unique name (no in-place-rename collapse).
- Task registers outputs from the validated `module.output_layer_indices` with a length-consistency guard.

**Backward-compat / byte-identical:** default exports (indices unset) add no Identity node and no initializer — unchanged graph.

**Tests added** (`TestGemma4OutputLayerIndices`): last-index emission, Identity-of-lm_head-input faithfulness, config-driven index scaling (`num_hidden_layers=3`), and rejection of multiple / non-last / duplicate / out-of-range / negative, plus `None`/empty backward-compat. Full `gemma4_test.py` (33) green; targeted gemma4+MoE build-graph suite (158 passed, 8 skipped). `ruff check` + `ruff format --check` clean.

Left unmerged for your review/admin-merge.
